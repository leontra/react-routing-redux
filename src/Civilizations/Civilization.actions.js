import { SPREAD_CIVILIZATIONS } from "../Constants/action_types";

export const spreadCivilizations = list => ({ 
    type: SPREAD_CIVILIZATIONS, 
    payload: {
        list: list
    } 
});
