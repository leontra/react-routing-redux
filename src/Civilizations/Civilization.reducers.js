import { SPREAD_CIVILIZATIONS, RESET_STATE } from './../Constants/action_types';

const initialState = {
    civilizations: []
};

const CivilizationReducer = (state = initialState, action) => {
    switch (action.type) {
        case SPREAD_CIVILIZATIONS: 
            return Object.assign(
                {}, 
                state, 
                { civilizations: state.civilizations.concat(action.payload.list) }
            );
        case RESET_STATE: 
            return Object.assign( {}, initialState );
        default:
            return state;
    }
}

export default CivilizationReducer;
