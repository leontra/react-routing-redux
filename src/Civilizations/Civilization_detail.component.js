import React, { Component } from 'react';
import { connect } from 'react-redux';
import Presenter from './Civilizations.presenter';
import ScreenLog from './../ScreenLog/screen-log.component';
// import { spreadCivilizations } from './Civilization.actions';

// function mapDispatchToProps(dispatch) {
//   return {
//     addArticle: article => dispatch(addArticle(article))
//   };
// }

class CivilizationDetail extends Component {

  constructor(props) {
    super(props);
    this.state = {
      civilization: {},
      error: false,
      logMsg: "Loading..."
    };
    this.getListCivilizations();
  }

  getListCivilizations() {
    this.presenter = new Presenter();
    this.presenter.getCivilizations(this.OnError.bind(this))
    .then(this.OnGetList.bind(this))
    .catch(_error => {
      console.log("Get Civilizations Error");
    })
  }

  OnGetList(list) {
    const element = list.find( (obj) => {
      if (!this.props.match) return false;
      const { params } = this.props.match;
      return params.name === obj.name;
    });
    if (element) {
      this.getDetail(element.id);
    }
  }

  OnError(err) {
    this.setState({
      error: true,
      logMsg: String(err.msg)
    });
  }

  getDetail(id) {
    this.presenter = new Presenter();
    this.presenter.getCivilizationDetail(id, this.OnError.bind(this))
    .then(this.OnGetDetail.bind(this))
    .catch(_error => {
      console.log("Get Cilization Detail Error");
    })
  }

  OnGetDetail(obj) {
    this.setState({
      civilization: obj,
      logMsg: "Success"
    });
  }

  render() {
    return (
      <div>
        <h1>Civilization Detail</h1>
        <div>
          <p className="civilization-name">{this.state.civilization.name}</p>
          <p className="civilization-expansion">{this.state.civilization.expansion}</p>
          <p className="civilization-army-type">{this.state.civilization.army_type}</p>
          <p className="civilization-team-bonus">{this.state.civilization.team_bonus}</p>
        </div>
      </div>
    );
  }
}

export default CivilizationDetail;