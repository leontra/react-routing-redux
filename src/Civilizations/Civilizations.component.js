import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Presenter from './Civilizations.presenter';
import ScreenLog from './../ScreenLog/screen-log.component';
import ListItem from './../UIElements/list-item.element';
// import { withRouter } from 'react-router-dom';

class Civilizations extends Component {

  constructor(props) {
    super(props);
    this.state = {
      list: [],
      error: false,
      logMsg: "Loading..."
    };

    this.getListCivilizations();
  }

  getListCivilizations() {
    this.presenter = new Presenter();
    this.presenter.getCivilizations(this.OnError.bind(this))
    .then(this.OnGetList.bind(this))
    .catch(_error => {
      console.log("Get Civilizations Error");
    });
  }

  OnError(err) {
    this.setState({
      error: true,
      logMsg: String(err.msg)
    });
  }

  OnGetList(list) {
      this.setState({
        list: list,
        logMsg: "Success"
      });
  }

  OnDetail(index) {
    this.props.history.push('/civilizations/' + this.state.list[index].name);
    // this.props.history.push('/units');
  }

  render() {
    const Elements = this.state.list.map(
      (value, index) => (<ListItem key={"civ_" + index + "_" + value.name} name={value.name} OnClick={this.OnDetail.bind(this, [index])} />)
    );

    return (
      <div>
        <h1>Civilizations</h1>
        <ScreenLog msg={this.state.logMsg} />
        <div className="list-container">
          {Elements}
        </div>
      </div>
    );
  }
}

Civilizations.propTypes = {
  list: PropTypes.array
}

export default Civilizations;