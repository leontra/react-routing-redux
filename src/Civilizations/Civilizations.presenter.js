import FetchData from './../FetchData/fetch-data'; 
import API_DICTIONARY from "./../Api/api-dictionary";

class CivilizationsPresenter {

    async getCivilizations(DisplayError) {
        const data = await FetchData.get(API_DICTIONARY.CIVILIZATIONS, []);
        if (data && typeof data.civilizations !== 'undefined') {
            return data.civilizations;
        }
        
        DisplayError(data);
        return [];
    }

    async getCivilizationDetail(id, DisplayError) {
        const data = await FetchData.get(API_DICTIONARY.CIVILIZATION_DETAIL, {id: id});
        if (data) {
            return data;
        }
        
        DisplayError(data);
        return [];
    }
    
}

export default CivilizationsPresenter;