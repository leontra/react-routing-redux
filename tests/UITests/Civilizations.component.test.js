import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
// import ReactTestUtils from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';
import "core-js/stable";
import "regenerator-runtime/runtime";
import Civilizations from '../../src/Civilizations/Civilizations.component';

import Presenter from '../../src/Civilizations/Civilizations.presenter';
jest.mock('../../src/Civilizations/Civilizations.presenter', () => {
    return jest.fn().mockImplementation(() => {
        return {getCivilizations: () => {
            return new Promise((resolve, reject) => {
                reject();
            });
        }};
    });
});

beforeEach( () => {
    Presenter.mockClear();
});

it('renders Civilizations without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Civilizations list={[]} />, 
    div);
  ReactDOM.unmountComponentAtNode(div);
});

// it('matches the snapshot', () => {
//     const div = document.createElement('div');
//     ReactDOM.render(
//         <Civilizations list={[]} />, 
//         div
//     );
    
//     const tree = renderer
//     .create(<Civilizations list={[]} />)
//     .toJSON();
//     expect(tree).toMatchSnapshot();

//     ReactDOM.unmountComponentAtNode(div);
// });

it('does the update for civilization detail', () => {    
    let history = [];
    const civilizationComponent = mount(<Civilizations list={[]} history={history} />);
    civilizationComponent.setState( { list: [{name: 'test'}] } );
    civilizationComponent.find('.list-item-test').simulate('click');
    expect(history.length).toBe(1);
    expect(history[0]).toBe("/civilizations/test");
});
