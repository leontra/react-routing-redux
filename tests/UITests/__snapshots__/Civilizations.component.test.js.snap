// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
<div>
  <h1>
    Civilizations
  </h1>
  <div>
    <p>
      Loading...
    </p>
  </div>
  <div
    className="list-container"
  />
</div>
`;
